examples/vae_text/vae_train.py

```diff
@@ -0,0 +1,477 @@
+                    type=str,
+                    default=None,
+                    help="Model path for generating sentences.")
+parser.add_argument('--out',
+                    type=str,
+                    default=None,
+                    help="Generation output path.")
+
+args = parser.parse_args()
+
+
+def kl_divergence(means: Tensor, logvars: Tensor) -> Tensor:
+    """Compute the KL divergence between Gaussian distribution
+    """
+
+    kl_cost = -0.5 * (logvars - means**2 -
+                      torch.exp(logvars) + 1.0)
+    kl_cost = torch.mean(kl_cost, 0)
+
+    return torch.sum(kl_cost)
+
+
+class VAE(nn.Module):
+    _latent_z: Tensor
+
+    def __init__(self,
+                 vocab_size: int, config_model):
+        super().__init__()
+        # Model architecture
+        self._config = config_model
+        self.encoder_w_embedder = tx.modules.WordEmbedder(
+            vocab_size=vocab_size, hparams=config_model.enc_emb_hparams)
+
+        self.encoder = tx.modules.UnidirectionalRNNEncoder(
+            input_size=self.encoder_w_embedder.dim,
+            hparams={
+                "rnn_cell": config_model.enc_cell_hparams,
+            })
+
+        self.decoder_w_embedder = tx.modules.WordEmbedder(
+            vocab_size=vocab_size, hparams=config_model.dec_emb_hparams)
+
+        if config_model.decoder_type == "lstm":
+            self.decoder = tx.modules.BasicRNNDecoder(
+                input_size=(self.decoder_w_embedder.dim +
+                    config_model.batch_size),
+                vocab_size=vocab_size,
+                token_embedder=self._embed_fn_rnn,
+                hparams={"rnn_cell": config_model.dec_cell_hparams})
+            sum_state_size = self.decoder.cell.hidden_size * 2
+
+        elif config_model.decoder_type == 'transformer':
+            # position embedding
+            self.decoder_p_embedder = tx.modules.SinusoidsPositionEmbedder(
+                position_size=config_model.max_pos,
+                hparams=config_model.dec_pos_emb_hparams)
+            # decoder
+            self.decoder = tx.modules.TransformerDecoder(
+                # tie word embedding with output layer
+                output_layer=self.decoder_w_embedder.embedding,
+                token_pos_embedder=self._embed_fn_transformer,
+                hparams=config_model.trans_hparams)
+            sum_state_size = self._config.dec_emb_hparams["dim"]
+        else:
+            raise ValueError("Decoder type must be 'lstm' or 'transformer'")
+
+        self.connector_mlp = tx.modules.MLPTransformConnector(
+            config_model.latent_dims * 2,
+            linear_layer_dim=self.encoder.cell.hidden_size * 2)
+
+        self.mlp_linear_layer = nn.Linear(
+            config_model.latent_dims,
+            sum_state_size)
+
+    def forward(self, data_batch: tx.data.Batch,
+                kl_weight: float, start_tokens: torch.LongTensor,
+                end_token: int) \
+            -> Dict:
```

Comment:
Don't wrap when it's unnecessary. Also, let's not use incomplete types. In this case, the return type could be `Dict[str, torch.Tensor]`.

Reply:
Got that.